src/engine/ContextDependentPiecewiseLinearConstraint.h

```diff
@@ -0,0 +1,246 @@
+class ContextDependentPiecewiseLinearConstraint
+   : public virtual PiecewiseLinearConstraint
+{
+public:
+    ContextDependentPiecewiseLinearConstraint();
+    ContextDependentPiecewiseLinearConstraint( unsigned numCases );
+
+    virtual ~ContextDependentPiecewiseLinearConstraint()
+    {
+        cdoCleanup();
+    }
+
+    /*
+      Turn the constraint on/off.
+    */
+    void setActiveConstraint( bool active );
+    bool isActive() const;
+
+    /*
+       Returns a list of all cases of this constraint and is used by the
+       nextFeasibleCase during search. The order of returned cases affects the
+       search, and this method is where related heuristics should be
+       implemented.
+     */
+    virtual List<PhaseStatus> getAllCases() const = 0;
+
+    /*
+     * Returns case split corresponding to the given phase/id
+       TODO: Update the signature in PiecewiseLinearConstraint, once the new
+       search is integrated.
+     */
+    virtual PiecewiseLinearCaseSplit getCaseSplit( PhaseStatus caseId ) const = 0;
+
+
+    /*
+      Check if the constraint's phase has been fixed.
+    */
+    virtual bool phaseFixed() const = 0;
+
+    /*
+      If the constraint's phase has been fixed, get the (valid) case split.
+    */
+    virtual PiecewiseLinearCaseSplit getImpliedCaseSplit() const = 0;
+
+    /*
+      Register a bound manager. If a bound manager is registered,
+      this piecewise linear constraint will inform the tightener whenever
+      it discovers a tighter (entailed) bound.
+    */
+    void registerBoundManager( BoundManager *boundManager );
+
+    /*
+       Register context object. Necessary for lazy backtracking features - such
+       as _cdPhaseStatus and _activeStatus. Does not require initialization until
+       after pre-processing.
+     */
+    void initializeCDOs( CVC4::context::Context *context );
+
+    /*
+       Politely clean up allocated CDOs.
+     */
+    void cdoCleanup();
+
+    /*
+      Get the context object - debugging purposes only
+    */
+    CVC4::context::Context *getContext() const
+    {
+        return _context;
+    }
+
+    /*
+      Get the active status object - debugging purposes only
+    */
+    CVC4::context::CDO<bool> *getActiveStatusCDO() const
+    {
+        return _cdConstraintActive;
+    };
+
+    /*
+      Get the current phase status object - debugging purposes only
+    */
+    CVC4::context::CDO<PhaseStatus> *getPhaseStatusCDO() const
+    {
+        return _cdPhaseStatus;
+    }
+
+    /*
+      Get the infeasible cases object - debugging purposes only
+    */
+    CVC4::context::CDList<PhaseStatus> *getInfeasibleCasesCDList() const
+    {
+        return _cdInfeasibleCases;
+    }
+
+    /*
+       Mark that an exploredCase is infeasible, reducing the remaining search space.
+     */
+    void markInfeasible( PhaseStatus exploredCase );
+
+    /*
+      Returns phaseStatus representing next feasible case.
+      Returns PHASE_NOT_FIXED if no feasible case exists.
```

Comment:
Can we perhaps add a dedicated sentinel? CONSTRAINT_INFEASIBLE or something?

Reply:
I have not done that because it would make dealing with Max and Disjunction trickier, but I will see what I can do.

Reply:
I was overthinking things, I added a high-valued entry to PhaseStatus enumeration, and updated this method accordingly.